File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        vector<int>arr;
        ListNode* temp=head;
        while(temp!=nullptr){
            arr.push_back(temp->val);
            temp=temp->next;
        }
        vector<int>brr;
        for(int i=0;i<(arr.size()+1)/2;i++){
            brr.push_back(arr[i]);
            if(i!=arr.size()-1-i)  {brr.push_back(arr[arr.size()-1-i]);}
        }
        ListNode* t=head;
        int i=0;
        int j=brr.size()-1;
        while(t!=nullptr && i<=j){
            t->val=brr[i];
            t=t->next;
            i++;
        }
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        fast = head.next
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        second = slow.next
        slow.next = None
        first = head

        prev = None
        nxt = None
        curr = second
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        

        while first and prev:
            temp1 = first.next
            temp2 = prev.next
           
            first.next = prev
            if temp1 is None:
                break
            prev.next = temp1

            first = temp1
            prev = temp2
        
